### PR TITLE
Add `DataColumnSidecar` gossip topic and message handling

### PR DESCRIPTION
**Pull request by jimmygchen**
## Issue Addressed

Part of #6072. 

This PR contains:
- Add gossip data column topic and message handling (excludes gossip validation)

Upstream PRs:
- #5050

### CODE REVIEW

File: beacon_node/lighthouse_network/src/types/pubsub.rs
Comment (pawanjay176) [inline beacon_node/lighthouse_network/src/types/pubsub.rs:287]: This could be dangerous as the slot value can be set to anything which could lead to a column sidecar bubbling up the stack and hitting one of our `unimplemented!`.
This wouldn't happen because we wouldn't subscribe to the required topic but better to be safe at every point in the networking stack specially given that we want this code in stable.

Comment (pawanjay176) [reply beacon_node/lighthouse_network/src/types/pubsub.rs:287]: Also, curious why we have allow for both electra + deneb. Is it just to provide flexibility for devnets on top of deneb/electra or something else?

File: beacon_node/network/src/sync/manager.rs
Comment (pawanjay176) [inline beacon_node/network/src/sync/manager.rs]: can we not have a panic macro here and just the comment? Just trying to be extra careful as this will go to stable eventually.
A change in gossip_methods in a future PR without the corresponding change here might go unnoticed

File: consensus/types/src/data_column_subnet_id.rs
Comment (pawanjay176) [inline consensus/types/src/data_column_subnet_id.rs:23]: Curious why we don't enforce that `column_index < data_column_sidecar_subnet_count` instead of taking the remainder

File: consensus/types/src/data_column_subnet_id.rs
Comment (pawanjay176) [inline consensus/types/src/data_column_subnet_id.rs]: Maybe do something like 
```rust
let hash_prefix: [u8; 8] = hash[0..8].try_into().expect("hash_fixed produces a 32 byte array");
```

File: consensus/types/src/data_column_subnet_id.rs
Comment (pawanjay176) [inline consensus/types/src/data_column_subnet_id.rs:145]: might be good to have a link to the test vectors here

**Review by pawanjay176 — CHANGES_REQUESTED** [review]
Looks mostly good. The das logic seems sane, but I have mainly focused on ensuring that we don't trigger any of the das related code if this is run on mainnet.

Just to be extra safe, I would prefer to remove all instances of panic macros and just let the `TODO(das)` comments guide us to incomplete code.

Also confirmed that we are not subscribing to any data column subnet topics in this PR so any of the upstream code wouldn't get triggered

File: consensus/types/src/data_column_subnet_id.rs
Comment (jimmygchen) [reply consensus/types/src/data_column_subnet_id.rs:23]: Not sure if I understand. This function calculates the `DataColumnSubnetId` from a `ColumnIndex`. `column_index` isn't necessary smaller than `data_column_sidecar_subnet_count`. The current spec value for `data_column_sidecar_subnet_count` is `32`, and we've been doing interop testing  with `64`. Latest proposal is to bump this value further up to `128`. Essentially we need to be able to support different combinations as both `NUMBER_OF_COLUMNS` and `DATA_COLUMN_SUBNET_COUNT` are configurable values.

File: consensus/types/src/data_column_subnet_id.rs
Comment (jimmygchen) [reply consensus/types/src/data_column_subnet_id.rs:145]: These tests were written before the `consensus-spec` tests were added and the `node_ids` here were taken from the `subnet_id.rs` tests:
https://github.com/sigp/lighthouse/blob/c52c598f6922059900d60723bb541add6865a87e/consensus/types/src/subnet_id.rs#L181

I can remove these tests now they're covered by the spec tests? (will require updating to `v1.5.0-alpha.3` though)

Comment (jimmygchen) [reply consensus/types/src/data_column_subnet_id.rs:145]: A bit of a trade off here - it's kinda nice to have some basic unit tests next to the actual code, these are our own unit tests, whereas the spec tests are to make sure we interop with others - though having overlapping tests adds maintenance effort.

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/manager.rs]: Yeah agree! I'll remove it. 

I originally added it cause I thought that was the preference when we were working on Deneb, but I think that was also before merging to `unstable`.

File: beacon_node/lighthouse_network/src/types/pubsub.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/types/pubsub.rs:287]: I've removed the `unimplemented!` calls, if there anything that you think we should change here?

I actually added this extra check (not currently on `das` branch) to make sure we don't break Deneb and Electra, but you're right that the slot can be set to anything. Is it possible though given that `EIP7594_FORK_EPOCH` is configured to `u64::max` on mainnet? As you said we've got some protection as we'll only subscribe to data column subnets on `EIP7594_FORK_EPOCH`.

Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/types/pubsub.rs:287]: > Also, curious why we have allow for both electra + deneb. Is it just to provide flexibility for devnets on top of deneb/electra or something else?

Yes, we're currently testing on top of Deneb but will switch over to Electra once it's stable enough for us to rebase. I'll add a TODO comment here to make sure we don't forget changing this.

File: consensus/types/src/data_column_subnet_id.rs
Comment (jimmygchen) [reply consensus/types/src/data_column_subnet_id.rs]: Updated in https://github.com/sigp/lighthouse/pull/6147/commits/29aab7a11afd862b09291947698c017e088bc872

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/manager.rs]: Removed in https://github.com/sigp/lighthouse/pull/6147/commits/29aab7a11afd862b09291947698c017e088bc872

File: consensus/types/src/data_column_subnet_id.rs
Comment (pawanjay176) [reply consensus/types/src/data_column_subnet_id.rs:145]: Ohh nice, didn't realise these are from our own tests.
That's neat.

File: beacon_node/lighthouse_network/src/types/pubsub.rs
Comment (pawanjay176) [reply beacon_node/lighthouse_network/src/types/pubsub.rs:287]: Yeah went through the gossipsub code and can confirm that a gossipsub message will not reach the application (lighthouse) unless it is subscribed to it. So we are good here.

File: consensus/types/src/data_column_subnet_id.rs
Comment (pawanjay176) [reply consensus/types/src/data_column_subnet_id.rs:23]: Ohh I miunderstood this one, looks good then

**Review by pawanjay176 — APPROVED** [review]

File: beacon_node/lighthouse_network/src/types/pubsub.rs
Comment (dapplion) [inline beacon_node/lighthouse_network/src/types/pubsub.rs:279]: Use deneb_activated()

File: beacon_node/network/src/network_beacon_processor/gossip_methods.rs
Comment (dapplion) [inline beacon_node/network/src/network_beacon_processor/gossip_methods.rs:621]: This should be an observation of a histogram not a gauge, same for `BEACON_DATA_COLUMN_DELAY_GOSSIP_VERIFICATION`

File: beacon_node/network/src/sync/manager.rs
Comment (dapplion) [inline beacon_node/network/src/sync/manager.rs:111]: It's actually useless to have to forward the payload in this error :)
- https://github.com/sigp/lighthouse/pull/5735

File: beacon_node/network/src/network_beacon_processor/gossip_methods.rs
Comment (jimmygchen) [reply beacon_node/network/src/network_beacon_processor/gossip_methods.rs:621]: This was changed for block and blobs in #5510 and I made the same change here for consistency.

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/manager.rs:111]: I think you're referring to `GossipDataColumnError::ParentUnknown` instead of `SyncMessage` here? we do need to keep the data column in the `BlockComponent` when doing a parent lookup.

File: beacon_node/lighthouse_network/src/types/pubsub.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/types/pubsub.rs:279]: Updated to use `deneb_enabled` in https://github.com/sigp/lighthouse/pull/6147/commits/f4cd1fa47c750a73b32760109e7a0ec91a1e0704

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [reply beacon_node/network/src/sync/manager.rs:111]: I've removed the unused error types in `GossipDataColumnError` and leaving this line unchanged. Let me know if you still want the payload removed for this error?

File: beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs
Comment (michaelsproul) [inline beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs]: We can do this later, but we should probably avoid cloning the `ChainSpec` so frequently by `Arc`ing it

**Review by michaelsproul — APPROVED** [review]
Approved modulo the `ChainSpec` cloning in `subnet_predicate` which could represent a performance regression on mainnet

File: beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/discovery/subnet_predicate.rs]: To do it properly we should probably `Arc` it everywhere, but we could probably do it in a separate PR as it will be  quite a big change. I've `Arc`'d the spec only in `Discovery` and `NetworkService` to avoid the performance regression for now.

**Review by michaelsproul — APPROVED** [review]

**Comment by michaelsproul** [comment]
@mergify queue

**Pull request merged**
